openmc/deplete/microxs.py

```diff
@@ -0,0 +1,229 @@
+            else:
+                xs[rxn] = ArbitraryXS(rxn, domain=reaction_domain, groups=groups, by_nuclide=True)
+            tallies += xs[rxn].tallies.values()
+
+        model.tallies = tallies
+
+        # create temporary run
+        with tempfile.TemporaryDirectory() as temp_dir:
+            statepoint_path = model.run(cwd=temp_dir)
+
+            with StatePoint(statepoint_path) as sp:
+                for rxn in xs:
+                    xs[rxn].load_from_statepoint(sp)
+
+        # Build the DataFrame
+        micro_xs = cls()
+        for rxn in xs:
+            df = xs[rxn].get_pandas_dataframe(xs_type='micro')
+            df.index = df['nuclide']
+            df.drop(['nuclide', xs[rxn].domain_type, 'group in', 'std. dev.'], axis=1, inplace=True)
+            df.rename({'mean': rxn}, axis=1, inplace=True)
+            micro_xs = concat([micro_xs, df], axis=1)
+
+        # Revert to the original tallies
+        model.tallies = original_tallies
+
+        return micro_xs
+
+    @classmethod
+    def _add_dilute_nuclides(cls, chain_file, model, dilute_initial):
+        chain = Chain.from_xml(chain_file)
+        reactions = chain.reactions
+        cross_sections = _find_cross_sections(model)
+        nuclides_with_data = cls._get_nuclides_with_data(cross_sections)
+        burnable_nucs = [nuc.name for nuc in chain.nuclides
+                               if nuc.name in nuclides_with_data]
+        diluted_materials = Materials()
+        for material in model.materials:
+            if material.depletable:
+                nuc_densities = material.get_nuclide_atom_densities()
+                dilute_density = 1.E-24 * dilute_initial
+                material.set_density('sum')
+                for nuc, density in nuc_densities.items():
+                    material.remove_nuclide(nuc)
+                    material.add_nuclide(nuc, density)
+                for burn_nuc in burnable_nucs:
+                    if burn_nuc not in nuc_densities:
+                        material.add_nuclide(burn_nuc,
+                                             dilute_density)
+            diluted_materials.append(material)
+        return reactions, diluted_materials
+
+    @staticmethod
+    def _get_nuclides_with_data(cross_sections):
+        """Loads cross_sections.xml file to find nuclides with neutron data
+
+        Parameters
+        ----------
+        cross_sections : str
+            Path to cross_sections.xml file
+
+        Returns
+        -------
+        nuclides : set of str
+            Set of nuclide names that have cross secton data
+
+        """
+        nuclides = set()
+        data_lib = DataLibrary.from_xml(cross_sections)
+        for library in data_lib.libraries:
+            if library['type'] != 'neutron':
+                continue
+            for name in library['materials']:
+                if name not in nuclides:
+                    nuclides.add(name)
+
+        return nuclides
```

Comment:
This is exactly the same as `CoupledOperator._get_nuclides_with_data`. Let's just make this a normal function and then call it from both the `CoupledOperator` and `MicroXS` classes.

Reply:
done